server/src/main/java/org/opensearch/common/util/FastLongHash.java

```diff
@@ -0,0 +1,307 @@
+
+        // The existing value at idx.
+        long existingValue;
+
+        // Always set the newly inserted key at its ideal home slot, even if it doesn't conform
+        // to the RHH scheme (yet). This will ensure subsequent correlated lookups are fast due
+        // to no additional probing. When another insertion causes this value to be displaced, it
+        // will eventually be placed at an appropriate location defined by the RHH scheme.
+        if ((value = table.set(idx, value)) == -1) {
+            // The ideal home slot was already empty; append the key and return early.
+            return append(key);
+        }
+
+        // Find an alternative slot for the displaced value such that the longest PSL is minimized.
+        do {
+            idx = (idx + 1) & mask;
+            value += INCR_PSL;
+
+            if ((existingValue = table.get(idx)) == -1) {
+                // Empty slot; insert the candidate value here.
+                table.set(idx, value);
+                return append(key);
+            } else if ((existingValue & MASK_PSL) <= (value & MASK_PSL)) {
+                // Existing value is "richer" than the candidate value at this index;
+                // swap and find an alternative slot for the displaced value.
+                // In the case of a tie, the candidate value (i.e. the recent value) is chosen as
+                // the winner and kept closer to its ideal home slot in order to speed up
+                // correlated lookups.
+                value = table.set(idx, value);
+            }
+        } while (true);
+    }
+
+    /**
+     * Appends the key in the keys' table.
+     */
+    private long append(final long key) {
+        keys = bigArrays.grow(keys, size + 1);
+        keys.set(size, key);
+        return size++;
+    }
+
+    /**
+     * Returns the hash for the given key.
+     * Visible for unit-tests.
+     */
+    long hash(final long key) {
+        return BitMixer.mix64(key);
+    }
+
+    /**
+     * Returns the underlying hash table.
+     * Visible for unit-tests.
+     */
+    LongArray getTable() {
+        return table;
+    }
+
+    /**
+     * Grows the hash table by doubling its capacity and reinserting the keys.
+     */
+    private void grow() {
+        // Ensure that the hash table doesn't grow too large.
+        // This implicitly also ensures that the ordinals are no larger than 2^32, thus,
+        // preventing them from polluting other bits (PSL/fingerprint) in the hash table values.
+        assert capacity < MAX_CAPACITY : "hash table already at the max capacity";
+
+        final long oldSize = size;
+        capacity <<= 1;
+        mask = capacity - 1;
+        size = 0;
+        grow = (long) (capacity * loadFactor);
+        table = bigArrays.resize(table, capacity);
+        table.fill(0, capacity, -1);
+
+        for (long ordinal = 0; ordinal < oldSize; ordinal++) {
+            insert(keys.get(ordinal));
+        }
```

Comment:
I considered this, but I didn't have sufficient data points to come up with a reasonable number. Moreover, I wanted to match the memory usage (initial capacity, load factor, and allocations) with the existing implementation.

Though, I'd like to point out that starting with a higher initial capacity may still be beneficial in terms of reducing heap allocations and GC pressure. The hash table is allocated using the `BigArrays` utility which allocates a primitive long array for size < 1024, but uses recycled arrays for size >= 1024. This means that resizing smaller tables requires allocation of new arrays (and GC of the old ones), which may be slower. Open for suggestions.

Reply:
I too don't have a good reasonable number to start with, but 1 seems too small. Would initial capacity of 32 or 64 work?

Reply:
I have updated the initial capacity to 32. Here's my rationale:
* Java hash maps have a default capacity of 16, so that's the least we should also have.
* Many aggregations would be like "24 hours of the day", "30 days of the month", etc. So having a capacity of 32 sounds reasonable. Unfortunately, I don't have data to prove this though. Queries like "60 minutes of the hour" or "60 seconds of the minute" might still require one resize.
* We also don't want to provision large tables and store only a few elements. Smaller tables are cache friendly as well.

Reply:
32 looks like a good start! Thanks!